ci(lv_templ): fix filename template mismatches and add CI

liamHowatt opened the pull request:
Fixes #8279

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

kisvegabor commented:
Wow, cool!

AndreCostaaa commented on .github/workflows/check_templ.yml:
```diff
@@ -0,0 +1,29 @@
+
+jobs:
+  template-check:
+    if: ${{ github.event_name != 'pull_request' || github.repository != github.event.pull_request.head.repo.full_name }}
```
I would vote to remove it. It doesn't add anything valuable and is hard to wrap our mind around it imo

It's basically saying "run if this is triggered by something other than a PR or if it's a PR from a fork"

liamHowatt replied:
I removed it. There are 12 CI workflows that have this check. I'd rather remove the rest in another PR so this one isn't "the one that broke all the CI".

vwheeler63 commented:
@liamHowatt  Please ignore the above regex for `.h` files -- testing is finding failures, so fixing that now.

vwheeler63 commented:
@liamHowatt Okay, the above regex for the `.h` files works now.  (Updated in above message.)

The one for `.c` files is
```
\s*/\*\*[\s\S]+?@file +(\S*)
```

I created a slightly more optimized script file that does the directory-tree walk and gets done in about 1 second on my medium-speed system.  It skips the `.git` and `docs` directories, and also skips `lv_font_...` files.

Be advised:

- Font files need to be read as a byte array and then decoded, or else the `f.read()` generates an exception (UnicodeDecodeError).
- I found there is no use to reading them since they are generated, and none match the pattern.

### Files that Perhaps Should Be in this PR (fixed):

- I found a number of The C-file "misses" (the regex doesn't match) that it seems should not be missed.
  - `./src/indev/lv_indev.c` due to `#include`s above the header comment, and quite a few `.h` files that don't have any comment after the ending `#endif`.
  - ./src/stdlib/builtin/lv_string_builtin.c  (comment after ending `#endif` makes no sense)
- I found a number of the H-file "misses" (the regex doesn't match) that it seems should not be missed.
  - ./src/misc/lv_area.h  (no comment after end `#endif`)
  - ./src/misc/lv_array.h  (no comment after end `#endif`)
  - ./src/misc/lv_math.h  (no comment after end `#endif`)
  - ./src/misc/lv_timer.h  (no comment after end `#endif`)
  - ./src/misc/lv_tree.h  (no comment after end `#endif`)
  - ./src/misc/lv_utils.h  (no comment after end `#endif`)
  - ./src/misc/cache/class/lv_cache_class.h  (`//` EOL comment after end `#endif`)
  - ./src/misc/cache/instance/lv_cache_instance.h  (`//` EOL comment after end `#endif`)
  - ./src/others/snapshot/lv_snapshot.h  (no comment after end `#endif`)

liamHowatt commented:
Thanks @vwheeler63 ! I will apply your regexs immediately if you can clarify this for me:

I looked up what `[\s\S]` does (a workaround for `.` not matching newlines), and I think `\s` is sufficient due to the `re.DOTALL` flag I'm using in `re.fullmatch` and `re.match`, which makes `.` match newlines. Is that right?

<hr>

> it will fail on Windows systems due to the enforced line endings

Python converts all line endings to `\n` on all platforms. The terminology is "universal line endings" in Python. Try it! See the `newline` parameter of `open` [here](https://docs.python.org/3/library/functions.html#open).

> If you encapsulate the initial assignment to `h_pattern` inside `re.compile(<string_here>)`, then it won't have to re-compile the regex each iteration through the loop. Same goes for any of the others that don't change.

The compiled regex are cached inside the regex module!

From [here](https://docs.python.org/3/library/re.html#re.compile):

> **Note:** The compiled versions of the most recent patterns passed to [re.compile()](https://docs.python.org/3/library/re.html#re.compile) and the module-level matching functions are cached, so programs that use only a few regular expressions at a time needn’t worry about compiling regular expressions.

Thus the need for [`re.purge()`](https://docs.python.org/3/library/re.html#re.purge)

> If you put the [recursive directory-tree walk](https://docs.python.org/3.12/library/os.html#os.walk) in the Python file, then someone could run it locally and get an instant read on whether he missed anything....  Reason:  not everyone has `find` on his platform.  I'd be happy to implement this after after this PR is merged if you'd like.

Ok I will add it! I don't trust the directory walker to not enter an inconsistent state if the files are modified while it's walking them, so I will collect them into a list first. This is the effect of `find` + `xargs`.

> We should make the filename after the `@file` command optional

Agreed.

> able to include a literal period in the name

Yes, that was intentional. Let me know how it can be better.

vwheeler63 commented:
@liamHowatt   A short note:  If it doesn't run for you in about 1 second, I would try pre-compiling the regex objects (which will require you to move the DOTALL flag to the compile step, instead of the search step).

liamHowatt commented:
I think I addressed everything! It's an overall improvement. The regex makes sense. Did I miss anything? Please review it again and let me know if I missed anything you said.

I see how reading as bytes and decoding makes it so original line endings are written back. Thanks for the Windows perspective.

The only thing I changed in your regex was `#endif */\* *([^\*]+) *\*/\s*` to `#endif */\* *([^* ]+) *\*/\s*`, See the `[\*]` part. It was causing the below change to be applied to 40 files.

```diff
-#endif /* LV_WAYLAND_SMM_H */
+#endif /* LV_WAYLAND_SMM_H*/
```

> Makes sense. The nice thing about os.walk() is that it is completely done processing the directory when it passes back the results.

Source? :sweat_smile: In the meantime I have `list(os.walk(project_dir))`

> I would try pre-compiling the regex objects (which will require you to move the DOTALL flag to the compile step, instead of the search step).

Brilliant. I like that.

> I found a number of The C-file "misses"

> I found a number of the H-file "misses"

These are still not addressed in this current state, FYI.

liamHowatt commented:
Forgot to add a ping: @vwheeler63

vwheeler63 commented:
@liamHowatt   Also -- an alternate way of opening files on Windows systems which forces `utf-8` decoding (and eliminates the `\r\n` sequence from opening it as a `bytes` array) is:

```python
open(fn, 'r', encoding='utf-8')
```
otherwise on Windows systems it tries to use an ANSI decoding.

liamHowatt commented:
> What I meant was that it fully finishes inside each directory before delivering the "root" and the 2 lists (files and dirs). To my knowledge, it does not do the whole tree walk first and only THEN return control, however I suspect your wrapping it in a `list()` function likely does cause that effect.

Ah okay. If you can share something to prove this I will remove `list()`.

Otherwise this is ready for review.

liamHowatt commented:
@vwheeler63 I turned the idea over in my head a bit and agree that it probably is safe without `list()` because of the way it returns full lists, as you said, so I removed it.

Ready for review.
